Show products information in analytics hub

rfermontero opened the pull request:
Closes: #5237

### Description
This pull request should close the analytics hub part 1. Here the new list card #5442 would be used to display the last card of the Hub.

### Testing instructions
There are unit tests for the information and to manual test please, follow this steps:

1. Open the application compiled in debug mode
2. Click on analytics section icon on bottom navigation bar
3. Scroll down to see the products card
4. Products card should appear at bottom

### Images/gif

<img src="https://user-images.githubusercontent.com/572601/145792113-852b5638-94a0-4fda-a01a-96832e49ba42.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792120-5c58196c-36aa-40d0-9783-d2de720196d4.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792122-bedcb8d6-523e-4a49-b310-1a5ab6496508.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792125-608c0e0c-1d57-4a78-ab3f-8255add4b715.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792128-04bb8210-13c6-4f4e-9586-0315c673d436.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792136-b4f6dddc-e3e1-4f19-a413-0c2855662561.png" width="400" height="1000">
<img src="https://user-images.githubusercontent.com/572601/145792640-e4a0c91b-5964-4438-93b9-7ff86bcdee77.png" width="400" height="1000">



- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rfermontero committed:
Get products revenue data in viewmodel

rfermontero committed:
Update analyticsProductCard information with viewState

rfermontero committed:
Render analytics list card delta tag

rfermontero committed:
Update products in view model

rfermontero committed:
Return itemsSold in analytics repository fetchProductStats

rfermontero committed:
Minor fixes in xml

rfermontero committed:
Add product list item description string

rfermontero committed:
Fetch and render products data in card

rfermontero committed:
Render list items and divider

rfermontero committed:
Create analytics list card item xml

rfermontero committed:
Minor fixes to analytics fragment

rfermontero committed:
Add skeleton for analytics list card

rfermontero committed:
Set specific message for no products card

rfermontero committed:
Add products stats model

rfermontero committed:
Implement fetchProductStats in repository and send data to view model

rfermontero committed:
Merge branch 'feature/show-analytics-orders-data' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Test view model for products card view state

rfermontero committed:
Merge branch 'feature/show-analytics-orders-data' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Merge branch 'feature/analytics-create-list-card' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Fix see report click

rfermontero committed:
Remove build.gradle uneeded change

rfermontero committed:
Merge branch 'feature/analytics-create-list-card' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Merge branch 'feature/analytics-create-list-card' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Fix tests

rfermontero committed:
Fix lint issue

rfermontero committed:
Merge branch 'develop' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Merge branch 'feature/analytics-hub' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Merge branch 'feature/analytics-hub' of github.com:colymore/woocommerce-android into feature/analytics-show-products-card

rfermontero committed:
Fix lint issues

rfermontero committed:
Usage of delta percentage in delta instead of Int

rfermontero committed:
Add tests for repository

rfermontero committed:
Hide delta tag when is NotExistant

rfermontero committed:
Minor adjust to right element of analytics list view

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -89,8 +91,56 @@ class AnalyticsRepository @Inject constructor(
+    ): Flow<ProductsResult> =
+        getGranularity(selectedRange).let { statsGranularity: StatsGranularity ->
+            return combine(
+                getCurrentPeriodStats(dateRange, statsGranularity),
+                getPreviousPeriodStats(dateRange, statsGranularity),
```
❓ Why are we fetching the revenue stats here again? Since it's for the same time period, can't we reuse the data we fetch for the revenue stats card?

rfermontero replied:
I'm doing this way because the HTTP call should be cached or saved in DB and if its not, I think that case would need storage for these stats which would be more error-prone I think. Do you see any approach that I'm not seeing?

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsViewModel.kt:
```diff
@@ -144,6 +162,33 @@ class AnalyticsViewModel @Inject constructor(
+                                it.productsStat.products
+                            )
+                        )
+                        ProductsError -> mutableState.value = state.value.copy(
+                            productsState = ProductsNoDataState(
+                                resourceProvider.getString(R.string.analytics_products_no_data)
+                            )
```
❓ Can we update the error handling for the cards so that when it's an error, we display a Snackbar that says error occurred, instead of just saying `No products` or `No orders` etc?

rfermontero replied:
Sure. In fact, I created an issue last week to define this because the designs are not completed for that case: #5439. As talked, makes sense to apply changes in another PR.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/listcard/AnalyticsListCardItemView.kt:
```diff
@@ -12,14 +16,19 @@ class AnalyticsListCardItemView @JvmOverloads constructor(
-        binding.analyticsCardListItemValue.text = viewState.title
+        binding.analyticsCardListItemValue.text = viewState.value
         binding.analyticsCardListItemDescription.text = viewState.description
+        binding.divider.visibility = if (viewState.showDivider == true) View.VISIBLE else View.GONE
```
np: we can simplify this logic to `binding.divider.isVisible = viewState.showDivider == true`

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/listcard/AnalyticsListCardItemViewState.kt:
```diff
@@ -1,8 +1,9 @@
+    val value: String,
     val description: String,
-    val value: Int,
+    val showDivider: Boolean? = true
```
np: are there instances when this `showDivider` can be null? If not, we can remove the `?` right? WDYT?

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/listcard/AnalyticsListViewState.kt:
```diff
@@ -8,6 +8,7 @@ sealed class AnalyticsListViewState {
         val subTitle: String,
         val subTitleValue: String,
         val delta: Int?,
+        val showDelta: Boolean,
```
np: Do you think we need another separate flag to know if we should show delta? Can't we check if `delta` is null?

rfermontero replied:
It could, of course, I made this way to make it consistent with the other 2 views, which have the same approach. I will change them.

anitaa1990 commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -283,13 +283,18 @@
+    <string name="analytics_products_card_title">Products</string>
+    <string name="analytics_products_list_header_title">Products</string>
+    <string name="analytics_products_list_header_subtitle">Items sold</string>
+    <string name="analytics_products_list_item_description">Net values: %1$s</string>
```
The design shows `Net Sales` here and not `Net values` 😅

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt:
```diff
@@ -545,6 +545,15 @@ class DateUtils @Inject constructor(
             null
         }
 
+    fun fromIso8601Format(date: String): Date? =
+        try {
+            SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault())
+                .parse(date)
+        } catch (e: Exception) {
+            "Error while parsing Iso8601 date to millis".reportAsError(e)
+            null
+        }
```
This method is already defined in `StringDateFormatExt` [here](https://github.com/woocommerce/woocommerce-android/blob/8df1568331e50ca5734a0b4db74c789156a9811d/WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/StringDateFormatExt.kt#L131). Can we reuse that?

rfermontero replied:
Op, I was looking for it. Changing.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/listcard/AnalyticsListCardView.kt:
```diff
@@ -32,30 +38,39 @@ class AnalyticsListCardView @JvmOverloads constructor(
+        viewState.delta?.let {
+            binding.analyticsItemsTag.tag = AnalyticsListDeltaTag(viewState.delta, getDeltaTagText(viewState))
+        }
+        binding.analyticsItemsList.removeAllViews()
```
❓ I'm curious why you didn't use a `RecyclerView` with adapter instead of inflating the list items manually 😄 We have a similar use case for the order detail screen and have the `OrderDetailProductListView` you can use for reference.

rfermontero replied:
I was thinking about it, but after thinking twice, it does not make sense. Changing to RV

anitaa1990 reviewed (commented):
- The data displayed for the same time period in the web and in the app is different for me (just for the products list).
- For yesterday, I have 2 products sold but this is how it is displayed in the web and in the app.
 <img src="https://user-images.githubusercontent.com/22608780/146323400-662b126b-4af5-4c52-93e6-29417cc1ac90.png" width="300"/>.  <img width="1494" alt="Screenshot 2021-12-16 at 11 19 17 AM" src="https://user-images.githubusercontent.com/22608780/146323491-15cdcdb6-72fe-4d66-8ad3-65eba223f432.png">

For Last month, the app shows 6 products sold but only displays 2.
<img src="https://user-images.githubusercontent.com/22608780/146323634-b6f2a2e9-2749-4e86-9f7a-e5a98788ffee.png" width="300"/>

I can add you to my test site if you would like to debug some more :) 

- The design specifies `Net sales` not `Net values` 😄 
- There are some tests failing in `GetTopPerformersTest`.

rfermontero committed:
Rename net values to net sales on analytics item element

rfermontero committed:
Simplify visibilities setter for elements

rfermontero committed:
Usage of already existant string ext method to get date

rfermontero committed:
Fix tests for get to performers

rfermontero committed:
Convert LinearLayout to recycleview

rfermontero committed:
Remove showDelta field, calculating with null check for delta value

rfermontero committed:
Fix checkstyle

rfermontero committed:
Remove shareIn

rfermontero committed:
Add mutex to getRevenue and implement force get from api decision

rfermontero committed:
Fix checkstyle issues

rfermontero committed:
Change query time range by start date and end date as iso format

rfermontero committed:
Remove fetchStrategy for products due to there is no way to filter locally

rfermontero commented:
> * The data displayed for the same time period in the web and in the app is different for me (just for the products list).
> * For yesterday, I have 2 products sold but this is how it is displayed in the web and in the app.
>   <img alt="" width="300" src="https://user-images.githubusercontent.com/22608780/146323400-662b126b-4af5-4c52-93e6-29417cc1ac90.png">.  <img alt="Screenshot 2021-12-16 at 11 19 17 AM" width="1494" src="https://user-images.githubusercontent.com/22608780/146323491-15cdcdb6-72fe-4d66-8ad3-65eba223f432.png">
> 
> For Last month, the app shows 6 products sold but only displays 2. <img alt="" width="300" src="https://user-images.githubusercontent.com/22608780/146323634-b6f2a2e9-2749-4e86-9f7a-e5a98788ffee.png">
> 
> I can add you to my test site if you would like to debug some more :)
> 
> * The design specifies `Net sales` not `Net values` 😄
> * There are some tests failing in `GetTopPerformersTest`.

I've fixed this, but I have needed a few commits to fix a little issue in FluxC, this is the PR (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2220) also I have changed a parameter here used to calculate dates for the request. Now it seems to be working fine.
After the FluxC pull request is merged I would need to change the fluxc version used in wc here.
Just mentioning that for this version, the number of values shown is fixed, do you think we should change to a higher number?

Exmaple of last month data shown:
![Screenshot 2021-12-16 at 19 53 02](https://user-images.githubusercontent.com/572601/146431424-4f5d7218-97ff-43b0-95af-d5c0060d31df.png)

rfermontero committed:
Usage of to date for SimpleDateRange when get products leader board

rfermontero committed:
Remove unused import

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -123,21 +226,50 @@ class AnalyticsRepository @Inject constructor(
+        endDate: String,
+        granularity: StatsGranularity,
+        fetchStrategy: FetchStrategy
+    ) = withContext(Dispatchers.IO) {
```
Since the StatsRepository is exposing a Flow, context switching should be done using the operator [`flowOn`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow-on.html), otherwise the execution will happen on the collector's dispatcher, which is `main` in our case, you can confirm this by putting a breakpoint [here](https://github.com/colymore/woocommerce-android/blob/35e90da4666ec7f0397ae5faa002436c0c29a2b7/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/data/StatsRepository.kt#L83) and check the thread.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -123,21 +226,50 @@ class AnalyticsRepository @Inject constructor(
+        endDate: String,
+        granularity: StatsGranularity,
+        quantity: Int,
+    ) = withContext(Dispatchers.IO) {
```
Same as above :)

rfermontero replied:
> Thanks for your work on this @colymore, nice job, I left some comments for your attention 🙂

It has been a pleasure :D Thanks for your comments too!

hichamboushaba commented on WooCommerce/src/main/res/layout/skeleton_analytics_list_card.xml:
```diff
@@ -0,0 +1,105 @@
+        app:layout_constraintTop_toBottomOf="@+id/leftVal" />
+
+    <include
+        android:id="@+id/list_item_1"
```
np, since you used camelcase for the ids of the views above, I think it would've been more consistent if you did the same here and below 🙂

rfermontero replied:
Totally right, I forgot to change this after reading the WC conventions regarding layout ids.
Changing.

hichamboushaba reviewed (commented):
Thanks for your work on this @colymore, nice job, I left some comments for your attention 🙂

rfermontero committed:
Convert fetch functions to regular instead of flow. Fix concurrency problems with a cache

rfermontero committed:
Rename to camel case skeleton list items

rfermontero committed:
Fix tests for analytics repository

rfermontero committed:
Create Analytics storage

rfermontero committed:
Use analytics storage inside analytics repository

rfermontero committed:
Change corotuines scopes for view model and repository

rfermontero commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -25,14 +28,22 @@ class AnalyticsRepository @Inject constructor(
+        selectedRange: AnalyticTimePeriod,
+        fetchStrategy: FetchStrategy
+    ): RevenueResult =
+        coroutineScope {
```
I am not completely sure about this coroutineScope change here to be honest, what do you think @hichamboushaba ?

hichamboushaba replied:
Unless I'm missing something, `coroutineScope` doesn't have any effect here, as its [docs](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/) states, it's useful for structured concurrency, if we want to start some child coroutines and wait and combine them (using `launch` or `async` for example)

rfermontero replied:
That was what I think, removing it :). Thanks!

rfermontero commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -123,21 +225,52 @@ class AnalyticsRepository @Inject constructor(
+        endDate: String,
+        granularity: StatsGranularity,
+        fetchStrategy: FetchStrategy
+    ): Flow<Result<WCRevenueStatsModel?>> = withContext(dispatchers.io) {
```
This is my approach, using a mutex on top of the call to do the block mutual exclusive and get data from in-memory storage which only will provide N maximum hits before clearing it up. 
This is solving the problem so far, there are only 3 requests per screen load. The first and second ones with different dates for getting revenue and a third one for the products leader board.

hichamboushaba replied:
I tested the code, and it's working 👍.
But can you explain why you opted for this solution instead of using a shared Deferred object for handling the cache?

I see one drawback for your solution, there is a tight coupling between the number of requests in the `AnalyticsRepository` and the constant `MAX_STATS_CACHE_CONSUMPTIONS`, so if someone who is not very familiar with the code came to add a third function in `AnalyticsRepository` which uses the same source, but didn't know about this coupling, and they missed updating it, it would break the logic, WDYT about this? 🙂

rfermontero replied:
Thank for test :) and yes you are completely right, the code works but it's coupled with the constant in the other repository.
I did like this because I feel more comfortable working with a map, atomic counter, and a mutex because I was not 100% sure about how could I deal with a Deferred object. I was thinking also of exposing a method to change the N max number for hits in the storage cache from the client but it would probably be error-prone if someone does not know that is a needed step to get it working fine.

I will think twice about how can I implement it with a deferred object now that I know how the code should work at the end, no matter the implementation details and we can check if the approach with it seems valid.
Thanks for your points, very helpful for me.

hichamboushaba replied:
> I was thinking also of exposing a method to change the N max number for hits in the storage cache from the client but it would probably be error-prone if someone does not know that is a needed step to get it working fine.

It doesn't have to be a variable, but instead a constructor parameter for the `AnalyticsStorage`, this way it will be very clear that it's a mandatory thing to offer.

> I will think twice about how can I implement it with a deferred object now that I know how the code should work at the end, no matter the implementation details and we can check if the approach with it seems valid.

It doesn't have to be done using Deferred, I just wanted to understand if you experimented with it or not yet, anyway, if you need any input, feel free to ping me 🙂.

rfermontero replied:
I've implemented the approach using deferred (Thanks to your skeleton, of course). Regarding the locks, I've moved them as close as possible to the network call and seems to be working fine for all cases.
Thank for the idea!

rfermontero committed:
Fix checkstyle issue

rfermontero committed:
Change fluxC version

rfermontero committed:
Change fluxC version

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsViewModel.kt:
```diff
@@ -206,7 +263,7 @@ class AnalyticsViewModel @Inject constructor(
     private fun getAvailableDateRanges() = resourceProvider.getStringArray(R.array.date_range_selectors).asList()
     private fun getDefaultTimePeriod() = AnalyticTimePeriod.TODAY
-    private fun getDefaultDateRange() = SimpleDateRange(
+    private fun getDefaultDateRange(): AnalyticsDateRange = SimpleDateRange(
```
This should probably proxy the call to `analyticsDateRange.getAnalyticsDateRangeFrom` to make sure we use the same value whether it's the default one, or it's selecting the `TODAY` option.

And a question, do we need to save both values `AnalyticTimePeriod` and `AnalyticsDateRange`? from what I saw in the code, `AnalyticsDateRange` is always derived from `AnalyticTimePeriod`, and if this is true, then having a single source of truth (`AnalyticTimePeriod`) would be less error-prone.

rfermontero replied:
Good point the first suggestion :).

Regarding the second question, currently, we don't need but it seems one of the first future tasks for this screen will be to create a custom date range that does not fit with the time period so I think it would be totally needed in a near future:).

rfermontero committed:
Remove unused analytics storage

rfermontero committed:
Usage of deferred computation as cached value for AnalyticsRepository get revenue and test it

rfermontero committed:
Move revenue mutex to grain fine lock

rfermontero committed:
Usage of analytics date range function instead of precalculated value

rfermontero committed:
Fix checkstyle

rfermontero committed:
Fix checkstyle

rfermontero committed:
Fix checkstyle

rfermontero commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -25,89 +31,194 @@ class AnalyticsRepository @Inject constructor(
-                        )
-                    )
-                }
+        val resultsAreValid = validateRevenueResults(listOf(currentPeriodRevenue, previousPeriodRevenue)) {
```
I have wrapped the validations in a function like this to pass the checkstyle number of returns check

hichamboushaba replied:
Good job wrapping it, but I think we can do better, to avoid abusing the double-bang operator (I'm not saying the operator shouldn't be used, but there are better solutions for some cases).

Here, since we are already mapping the calls in `fetchNetworkStats`, then we can just add a single function call to make the response non-nullable, by adding this to the end of the function:
```kotlin
mapCatching { it!! }
```
This will make sure to convert our returned value from `Result<WCRevenueStatsModel?>` to `Result<WCRevenueStatsModel>` and it would return a failure result if the value is null.
Then we would be able to do something like this in our function:

```kotlin
suspend fun fetchRevenueData(
        dateRange: AnalyticsDateRange,
        selectedRange: AnalyticTimePeriod,
        fetchStrategy: FetchStrategy
    ): RevenueResult {
        val granularity = getGranularity(selectedRange)
        val currentPeriodRevenue = getCurrentPeriodStats(dateRange, granularity, fetchStrategy).getOrNull() ?: return RevenueError
        val previousPeriodRevenue =
            getPreviousPeriodStats(dateRange, granularity, fetchStrategy).getOrNull() ?: return RevenueError

        val previousTotalSales = previousPeriodRevenue.parseTotal()?.totalSales ?: 0.0
        val previousNetRevenue = previousPeriodRevenue.parseTotal()?.netRevenue ?: 0.0
        val currentTotalSales = currentPeriodRevenue.parseTotal()?.totalSales!!
        val currentNetRevenue = currentPeriodRevenue.parseTotal()?.netRevenue!!

        return RevenueData(
            RevenueStat(
                currentTotalSales,
                calculateDeltaPercentage(previousTotalSales, currentTotalSales),
                currentNetRevenue,
                calculateDeltaPercentage(previousNetRevenue, currentNetRevenue),
                getCurrencyCode()
            )
        )
    }
```

WDYT about a change like this? 🙂

rfermontero replied:
Very much simpler and it makes sense to apply it as soon as possible in the stack calls.

Anyway, I will need to do all the checks outside that lines but inside an if, to be able to have only 2 returns inside the functions, otherwise ReturnCount checkstyle exception will be raised.

rfermontero committed:
Fix test for AnalyticsViewModel

rfermontero commented:
@hichamboushaba When you have free time, could you take a look at the solution? :D Thanks!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -158,45 +152,61 @@ class AnalyticsRepository @Inject constructor(
+                AnalyticsStatsResultWrapper(
+                    startDate = startDate,
+                    endDate = endDate,
+                    result = getCurrentRevenueMutex.withLock {
+                        async { fetchNetworkStats(startDate, endDate, granularity, fetchStrategy) }
+                    }
```
`currentRevenueStats` is mutable, so if there were two concurrent calls, they would probably change it concurrently here, since the lock is only wrapping the `async` bloc, and not the whole instantiation 🤔

rfermontero replied:
I got your point, totally right, lock should be over the var wrapping even the shouldUpdate function and not over the operation.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -224,37 +234,45 @@ class AnalyticsRepository @Inject constructor(
 
-    private suspend fun fetchStats(
+    private fun shouldUpdatePreviousStats(startDate: String, endDate: String) =
+        (previousRevenueStats == null || previousRevenueStats?.result?.isCompleted == true) &&
```
I see you are updating the result every time it's completed, can you explain why?

From my perspective, unless there was a force refresh, even if the result is `completed`, it's still a valid one, and there is no need to go down the layers, WDYT?

rfermontero replied:
I was thinking about that for a while, let me explain my thoughts:

Force refresh is received in the repository as true when a user does a pull to refresh action. So, each time user does it I would receive 3 requests (One per card, so, Revenue, Products, and Orders) to force update data, If I would be using the force=true to decide when should I update the result, each card request would go through the network, even being the same.

With this flow, the force refresh is being sent to fluxC request payload only for the first call and the other 2 are waiting until it finishes and the result is received, so, it's avoiding those 2 last calls. Next time a force request is done, only the first call will be over the network, updating the results, and so on...

I'm adding a gif about the network calls done to verify only 3 record calls should be done when the date selector changes, one per revenue data and another one for products.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/572601/147090933-4707419d-20ca-4f4a-9cfe-8b096e9a6fec.gif)

So, in theory, this is working as expected.
Does it make sense for you?

hichamboushaba replied:
I didn't mean check only the `force` value, but combine with `isCompleted` to decide.
But actually I didn't check your code very well, it has an issue, since you have an `&&` operator there, unless we change the `date`, when we try to refresh, we will keep using the cached value.

I think a bit of code would make what I meant clearer:

```kotlin
private fun shouldUpdateCurrentStats(force: Boolean, startDate: String, endDate: String) =
        currentRevenueStats?.startDate != startDate || currentRevenueStats?.endDate != endDate
             || (force && currentRevenueStats?.result?.isCompleted == true)
```

If `currentRevenueStats` is null, then the first check will be `true`, but if itsn't, then we will check the dates first, and if they are the same, we won't refresh the result unless we need to force refresh, and with the `isCompleted` check, we can be sure we won't invalidate it while it's still running.

rfermontero replied:
I still don't see the point about the force there I will test it to understand it completely. 
Thanks for the suggestion!

rfermontero replied:
I understand your point now with the debugger and stopping in the calls. 
It seems clearest and safest.

hichamboushaba reviewed (commented):
Thanks @colymore for implementing this using the Deferred job, I left some comments on the PR, but nothing major.

rfermontero committed:
Move analytics current and previous period revenue lock over the check function.
Do MapCatching for revenue network calls and parse total on it to avoid nullable chain calls.

rfermontero commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -25,88 +31,191 @@ class AnalyticsRepository @Inject constructor(
+        val currentPeriodTotalRevenue = getCurrentPeriodStats(dateRange, granularity, fetchStrategy).getOrNull()
+        val previousPeriodTotalRevenue = getPreviousPeriodStats(dateRange, granularity, fetchStrategy).getOrNull()
+
+        if (listOf(currentPeriodTotalRevenue, previousPeriodTotalRevenue).any { it == null } ||
+            currentPeriodTotalRevenue?.totalSales == null ||
+            currentPeriodTotalRevenue.netRevenue == null
+        ) {
```
I need to do the checks here to get only one return XXXError, otherwise, checkstyle will fail because of too much returns in one function

rfermontero commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -25,88 +31,191 @@ class AnalyticsRepository @Inject constructor(
-                fetchStats(dateRange.from.formatToYYYYmmDD(), dateRange.from.formatToYYYYmmDD(), granularity)
-            is MultipleDateRange ->
-                fetchStats(dateRange.from.from.formatToYYYYmmDD(), dateRange.from.to.formatToYYYYmmDD(), granularity)
+        getCurrentRevenueMutex.withLock {
```
Lock over the first usage of the object result instead of the update. This should prevent concurrent updates having the mutual exclusion zone.

rfermontero committed:
Usage of force boolean parameter to decide if refresh cached revenue value

rfermontero committed:
Update tests for analytics repository concurrency fetch revenue

rfermontero committed:
Fix checkstyle

rfermontero commented:
Hi @hichamboushaba How are you doing? Please, when you have free time, take a look again on the latest changes, I've applied your suggestions :D.
Thanks!

hichamboushaba commented:
The latest changes looks good @colymore, thanks 👍

anitaa1990 commented:
Merging this PR to the feature branch for now 👍
